tensorly/decomposition/constrained_parafac.py

```diff
@@ -0,0 +1,284 @@
+                weights, factors = CPTensor((None, factors))
+            else:
+                weights_avg = tl.prod(weights) ** (1.0 / tl.shape(weights)[0])
+                for i in range(len(factors)):
+                    factors[i] = factors[i] * weights_avg
+            kt = CPTensor((None, factors))
+            return kt
+        except ValueError:
+            raise ValueError(
+                'If initialization method is a mapping, then it must '
+                'be possible to convert it to a CPTensor instance'
+            )
+    else:
+        raise ValueError('Initialization method "{}" not recognized'.format(init))
+
+    for i in range(n_modes):
+        factors[i] = proximal_operator(factors[i], constraint=constraints[i], reg_par=reg_par[i], prox_par=prox_par[i])
+    kt = CPTensor((None, factors))
+    return kt
+
+
+def constrained_parafac(tensor, rank, n_iter_max=100, n_iter_max_inner=10,
+                        init='svd', svd='numpy_svd',
+                        tol_outer=1e-8, tol_inner=1e-6, random_state=None,
+                        verbose=0, return_errors=False,
+                        constraints=None,
+                        reg_par=None,
+                        prox_par=None,
+                        cvg_criterion='abs_rec_error',
+                        fixed_modes=None):
+    """CANDECOMP/PARAFAC decomposition via alternating optimization of
+    alternating direction method of multipliers (AO-ADMM):
+
+    Computes a rank-`rank` decomposition of `tensor` [1]_ such that::
+        tensor = [|weights; factors[0], ..., factors[-1] |],
+    where factors are either penalized or constrained according to the user-defined constraint.
+
+    In order to compute the factors efficiently, the ADMM algorithm
+    introduces an auxilliary factor which is called factor_aux in the function.
+
+    Parameters
+    ----------
+    tensor : ndarray
+    rank  : int
+        Number of components.
+    n_iter_max : int
+        Maximum number of iteration for outer loop
+    n_iter_max_inner : int
+        Number of iteration for inner loop
+    init : {'svd', 'random', cptensor}, optional
+        Type of factor matrix initialization. See `initialize_factors`.
+    svd : str, default is 'numpy_svd'
+        function to use to compute the SVD, acceptable values in tensorly.SVD_FUNS
+    tol_outer : float, optional
+        (Default: 1e-8) Relative reconstruction error tolerance for outer loop. The
+        algorithm is considered to have found a local minimum when the
+        reconstruction error is less than `tol_outer`.
+    tol_inner : float, optional
+        (Default: 1e-6) Absolute reconstruction error tolerance for factor update during inner loop, i.e. ADMM optimization.
+    random_state : {None, int, np.random.RandomState}
+    verbose : int, optional
+        Level of verbosity
+    return_errors : bool, optional
+        Activate return of iteration errors
+    constraints : string or list of strings, optional
```

Comment:
I see the flexibility of this API but also find it slightly confusing: how about using an explicit parameters for each? 
e.g. non_negative=False, l1_reg=None, l2_reg=None, etc? 

This will be more verbose but perhaps clearer for users?